File: backend/app/schemas/card_pack.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.clothing_item import Tag
from app.schemas.creator import Pagination


class CardPackCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=160)
    description: str | None = None
    pack_type: Literal["CLOTHING_COLLECTION"] = Field(
        default="CLOTHING_COLLECTION",
        alias="type",
    )
    item_ids: list[UUID] = Field(alias="itemIds", min_length=1)
    cover_image: str | None = Field(default=None, alias="coverImage")

    @field_validator("item_ids")
    @classmethod
    def _validate_unique_item_ids(cls, value: list[UUID]) -> list[UUID]:
        if len(value) != len(set(value)):
            raise ValueError("itemIds must be unique")
        return value


class CardPackUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, max_length=160)
    description: str | None = None
    item_ids: list[UUID] | None = Field(default=None, alias="itemIds")
    cover_image: str | None = Field(default=None, alias="coverImage")

    @field_validator("item_ids")
    @classmethod
    def _validate_unique_item_ids(cls, value: list[UUID] | None) -> list[UUID] | None:
        if value is not None and len(value) != len(set(value)):
            raise ValueError("itemIds must be unique")
        return value


class CardPackItemSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    creator_item_id: UUID = Field(alias="creatorItemId")
    sort_order: int = Field(alias="sortOrder")
    name: str | None = None
    description: str | None = None
    catalog_visibility: str = Field(alias="catalogVisibility")
    processing_status: str = Field(alias="processingStatus")
    cover_url: str | None = Field(default=None, alias="coverUrl")
    final_tags: list[Tag] = Field(default_factory=list, alias="finalTags")
    created_at: datetime = Field(alias="createdAt")


class CardPackBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    creator_id: UUID = Field(alias="creatorId")
    creator_uid: str | None = Field(default=None, alias="creatorUid")
    creator_username: str | None = Field(default=None, alias="creatorUsername")
    name: str
    description: str | None = None
    pack_type: str = Field(alias="type")
    status: str
    cover_image: str | None = Field(default=None, alias="coverImage")
    wardrobe_id: UUID | None = Field(default=None, alias="wardrobeId")
    wardrobe_wid: str | None = Field(default=None, alias="wardrobeWid")
    share_id: str | None = Field(default=None, alias="shareId")
    import_count: int = Field(alias="importCount")
    published_at: datetime | None = Field(default=None, alias="publishedAt")
    archived_at: datetime | None = Field(default=None, alias="archivedAt")
    item_count: int = Field(alias="itemCount")
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")


class CardPackListItem(CardPackBase):
    pass


class CardPackDetail(CardPackBase):
    items: list[CardPackItemSummary] = Field(default_factory=list)


class CardPackPublishData(CardPackDetail):
    share_link: str = Field(alias="shareLink")


class CardPackListData(BaseModel):
    items: list[CardPackListItem]
    pagination: Pagination


class CardPackDetailResponse(BaseModel):
    success: bool = True
    data: CardPackDetail


class CardPackListResponse(BaseModel):
    success: bool = True
    data: CardPackListData


class CardPackPublishResponse(BaseModel):
    success: bool = True
    data: CardPackPublishData

File: ai_wardrobe_app/lib/models/card_pack.dart
enum PackType { clothingCollection, outfit }

enum PackStatus { draft, published, archived }

class CardPack {
  final String id;
  final String creatorId;
  final String? creatorUid;
  final String? creatorUsername;
  final String name;
  final String? description;
  final String? coverImageUrl;
  final String? wardrobeId;
  final String? wardrobeWid;
  final PackType type;
  final List<String> itemIds;
  final int itemCount;
  final String? shareLink;
  final PackStatus status;
  final int importCount;
  final DateTime createdAt;
  final DateTime? publishedAt;
  final DateTime updatedAt;
  final List<Map<String, dynamic>>? items;

  CardPack({
    required this.id,
    required this.creatorId,
    this.creatorUid,
    this.creatorUsername,
    required this.name,
    this.description,
    this.coverImageUrl,
    this.wardrobeId,
    this.wardrobeWid,
    required this.type,
    required this.itemIds,
    required this.itemCount,
    this.shareLink,
    required this.status,
    required this.importCount,
    required this.createdAt,
    this.publishedAt,
    required this.updatedAt,
    this.items,
  });

  factory CardPack.fromJson(Map<String, dynamic> json) {
    return CardPack(
      id: json['id'] as String,
      creatorId: json['creatorId'] as String,
      creatorUid: json['creatorUid'] as String?,
      creatorUsername: json['creatorUsername'] as String?,
      name: json['name'] as String,
      description: json['description'] as String?,
      coverImageUrl:
          json['coverImageUrl'] as String? ?? json['coverImage'] as String?,
      wardrobeId: json['wardrobeId'] as String?,
      wardrobeWid: json['wardrobeWid'] as String?,
      type: _parsePackType(json['type'] as String? ?? 'CLOTHING_COLLECTION'),
      itemIds: (json['itemIds'] as List<dynamic>? ?? const [])
          .map((e) => e.toString())
          .toList(),
      itemCount:
          json['itemCount'] as int? ??
          (json['itemIds'] as List<dynamic>? ?? const []).length,
      shareLink: json['shareLink'] as String?,
      status: _parsePackStatus(json['status'] as String? ?? 'DRAFT'),
      importCount: json['importCount'] as int? ?? 0,
      createdAt: DateTime.parse(
        json['createdAt'] as String? ?? DateTime.now().toIso8601String(),
      ),
      publishedAt: json['publishedAt'] != null
          ? DateTime.parse(json['publishedAt'] as String)
          : null,
      updatedAt: DateTime.parse(
        json['updatedAt'] as String? ?? DateTime.now().toIso8601String(),
      ),
      items: json['items'] != null
          ? (json['items'] as List<dynamic>)
                .map((e) => Map<String, dynamic>.from(e as Map))
                .toList()
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'creatorId': creatorId,
      'creatorUid': creatorUid,
      'creatorUsername': creatorUsername,
      'name': name,
      'description': description,
      'coverImageUrl': coverImageUrl,
      'wardrobeId': wardrobeId,
      'wardrobeWid': wardrobeWid,
      'type': switch (type) {
        PackType.clothingCollection => 'CLOTHING_COLLECTION',
        PackType.outfit => 'OUTFIT',
      },
      'itemIds': itemIds,
      'itemCount': itemCount,
      'shareLink': shareLink,
      'status': switch (status) {
        PackStatus.draft => 'DRAFT',
        PackStatus.published => 'PUBLISHED',
        PackStatus.archived => 'ARCHIVED',
      },
      'importCount': importCount,
      'createdAt': createdAt.toIso8601String(),
      'publishedAt': publishedAt?.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'items': items,
    };
  }

  static PackType _parsePackType(String type) {
    switch (type.toUpperCase()) {
      case 'OUTFIT':
        return PackType.outfit;
      case 'CLOTHING_COLLECTION':
      default:
        return PackType.clothingCollection;
    }
  }

  static PackStatus _parsePackStatus(String status) {
    switch (status.toUpperCase()) {
      case 'PUBLISHED':
        return PackStatus.published;
      case 'ARCHIVED':
        return PackStatus.archived;
      case 'DRAFT':
      default:
        return PackStatus.draft;
    }
  }
}
